File: src/main/java/no/nav/sosialhjelp/soknad/web/rest/ressurser/personalia/KontonummerRessurs.java
//package no.nav.sosialhjelp.soknad.web.rest.ressurser.personalia;
//
//import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde;
//import no.nav.sbl.soknadsosialhjelp.soknad.personalia.JsonKontonummer;
//import no.nav.sbl.soknadsosialhjelp.soknad.personalia.JsonPersonalia;
//import no.nav.security.token.support.core.api.ProtectedWithClaims;
//import no.nav.sosialhjelp.metrics.aspects.Timed;
//import no.nav.sosialhjelp.soknad.business.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository;
//import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
//import no.nav.sosialhjelp.soknad.domain.model.oidc.SubjectHandler;
//import no.nav.sosialhjelp.soknad.personalia.kontonummer.KontonummerSystemdata;
//import no.nav.sosialhjelp.soknad.web.sikkerhet.Tilgangskontroll;
//import org.springframework.stereotype.Controller;
//
//import javax.ws.rs.GET;
//import javax.ws.rs.PUT;
//import javax.ws.rs.Path;
//import javax.ws.rs.PathParam;
//import javax.ws.rs.Produces;
//import javax.xml.bind.annotation.XmlAccessType;
//import javax.xml.bind.annotation.XmlAccessorType;
//
//import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
//import static no.nav.sosialhjelp.soknad.web.utils.Constants.CLAIM_ACR_LEVEL_4;
//import static no.nav.sosialhjelp.soknad.web.utils.Constants.SELVBETJENING;
//
//@Controller
//@ProtectedWithClaims(issuer = SELVBETJENING, claimMap = {CLAIM_ACR_LEVEL_4})
//@Path("/soknader/{behandlingsId}/personalia/kontonummer")
//@Timed
//@Produces(APPLICATION_JSON)
//public class KontonummerRessurs {
//
//    private final Tilgangskontroll tilgangskontroll;
//    private final SoknadUnderArbeidRepository soknadUnderArbeidRepository;
//    private final KontonummerSystemdata kontonummerSystemdata;
//
//    public KontonummerRessurs(Tilgangskontroll tilgangskontroll, SoknadUnderArbeidRepository soknadUnderArbeidRepository, KontonummerSystemdata kontonummerSystemdata) {
//        this.tilgangskontroll = tilgangskontroll;
//        this.soknadUnderArbeidRepository = soknadUnderArbeidRepository;
//        this.kontonummerSystemdata = kontonummerSystemdata;
//    }
//
//    @GET
//    public KontonummerFrontend hentKontonummer(@PathParam("behandlingsId") String behandlingsId) {
//        tilgangskontroll.verifiserAtBrukerHarTilgang();
//        String eier = SubjectHandler.getUserId();
//        SoknadUnderArbeid soknadUnderArbeid = soknadUnderArbeidRepository.hentSoknad(behandlingsId, eier);
//        JsonKontonummer kontonummer = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getKontonummer();
//        String systemverdi;
//        if (kontonummer.getKilde().equals(JsonKilde.SYSTEM)) {
//            systemverdi = kontonummer.getVerdi();
//        } else {
//            systemverdi = kontonummerSystemdata.innhentSystemverdiKontonummer(eier);
//        }
//
//        return new KontonummerFrontend()
//                .withBrukerdefinert(kontonummer.getKilde() == JsonKilde.BRUKER)
//                .withSystemverdi(systemverdi)
//                .withBrukerutfyltVerdi(kontonummer.getKilde() == JsonKilde.BRUKER ? kontonummer.getVerdi() : null)
//                .withHarIkkeKonto(kontonummer.getHarIkkeKonto());
//    }
//
//    @PUT
//    public void updateKontonummer(@PathParam("behandlingsId") String behandlingsId, KontonummerFrontend kontonummerFrontend) {
//        tilgangskontroll.verifiserAtBrukerKanEndreSoknad(behandlingsId);
//        String eier = SubjectHandler.getUserId();
//        SoknadUnderArbeid soknad = soknadUnderArbeidRepository.hentSoknad(behandlingsId, eier);
//        JsonPersonalia personalia = soknad.getJsonInternalSoknad().getSoknad().getData().getPersonalia();
//        JsonKontonummer kontonummer = personalia.getKontonummer();
//        if (kontonummerFrontend.brukerdefinert) {
//            kontonummer.setKilde(JsonKilde.BRUKER);
//            if ("".equals(kontonummerFrontend.brukerutfyltVerdi)) {
//                kontonummerFrontend.brukerutfyltVerdi = null;
//            }
//            kontonummer.setVerdi(kontonummerFrontend.brukerutfyltVerdi);
//            kontonummer.setHarIkkeKonto(kontonummerFrontend.harIkkeKonto);
//        } else if (kontonummer.getKilde() == JsonKilde.BRUKER) {
//            kontonummer.setKilde(JsonKilde.SYSTEM);
//            kontonummerSystemdata.updateSystemdataIn(soknad, "");
//            kontonummer.setHarIkkeKonto(null);
//        }
//        soknadUnderArbeidRepository.oppdaterSoknadsdata(soknad, eier);
//    }
//
//    @XmlAccessorType(XmlAccessType.FIELD)
//    public static final class KontonummerFrontend {
//        public boolean brukerdefinert;
//        public String systemverdi;
//        public String brukerutfyltVerdi;
//        public Boolean harIkkeKonto;
//
//        public KontonummerFrontend withBrukerdefinert(boolean brukerdefinert) {
//            this.brukerdefinert = brukerdefinert;
//            return this;
//        }
//
//        public KontonummerFrontend withSystemverdi(String systemverdi) {
//            this.systemverdi = systemverdi;
//            return this;
//        }
//
//        public KontonummerFrontend withBrukerutfyltVerdi(String brukerutfyltVerdi) {
//            this.brukerutfyltVerdi = brukerutfyltVerdi;
//            return this;
//        }
//
//        public KontonummerFrontend withHarIkkeKonto(Boolean harIkkeKonto) {
//            this.harIkkeKonto = harIkkeKonto;
//            return this;
//        }
//    }
//}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/personalia/kontonummer/KontonummerRessurs.kt
package no.nav.sosialhjelp.soknad.personalia.kontonummer

import io.swagger.v3.oas.annotations.media.Schema
import jakarta.validation.Valid
import jakarta.validation.constraints.Pattern
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde
import no.nav.sbl.soknadsosialhjelp.soknad.personalia.JsonKontonummer
import no.nav.sosialhjelp.soknad.app.annotation.ProtectionSelvbetjeningHigh
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository
import no.nav.sosialhjelp.soknad.tilgangskontroll.Tilgangskontroll
import no.nav.sosialhjelp.soknad.v2.shadow.V2ControllerAdapter
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import no.nav.sosialhjelp.soknad.app.subjecthandler.SubjectHandlerUtils.getUserIdFromToken as eier

@RestController
@ProtectionSelvbetjeningHigh
@RequestMapping("/soknader/{behandlingsId}/personalia/kontonummer")
class KontonummerRessurs(
    private val tilgangskontroll: Tilgangskontroll,
    private val soknadUnderArbeidRepository: SoknadUnderArbeidRepository,
    private val kontonummerService: KontonummerService,
    private val controllerAdapter: V2ControllerAdapter,
) {
    @GetMapping
    fun hentKontonummer(
        @PathVariable("behandlingsId") behandlingsId: String,
    ): KontonummerFrontend {
        tilgangskontroll.verifiserAtBrukerHarTilgang()
        val konto = loadKontonummer(behandlingsId)
        if (konto.kilde == JsonKilde.SYSTEM && konto.verdi == null) {
            val kontonummerRegister = kontonummerService.getKontonummer(eier())
            storeKontonummer(
                behandlingsId,
                JsonKontonummer().apply {
                    kilde = JsonKilde.SYSTEM
                    verdi = kontonummerRegister
                },
            )
        }
        return mapJsonToDto(loadKontonummer(behandlingsId))
    }

    @PutMapping
    fun updateKontonummer(
        @PathVariable("behandlingsId") behandlingsId: String,
        @RequestBody @Valid
        kontoDto: KontonummerInputDto,
    ): KontonummerFrontend {
        tilgangskontroll.verifiserAtBrukerKanEndreSoknad(behandlingsId)
        val kontoJson = mapInputToJson(kontoDto, kontonummerService.getKontonummer(eier()))
        storeKontonummer(behandlingsId, kontoJson)

        // NyModell
        controllerAdapter.updateKontonummer(behandlingsId, kontoDto)

        return mapJsonToDto(kontoJson)
    }

    @Schema(description = "Kontonummer for bruker - obs: PUT med (systemverdi !== null) vil nullstille brukerutfyltVerdi")
    data class KontonummerFrontend(
        @Schema(readOnly = true, description = "Kontonummer fra kontoregisteret")
        val systemverdi: String? = null,
        @Schema(nullable = true, description = "Kontonummer fra bruker")
        val brukerutfyltVerdi: String? = null,
        @Schema(nullable = false, description = "Bruker oppgir at de ikke har konto")
        val harIkkeKonto: Boolean = false,
    ) {
        @Deprecated("Unødvendig, utled av annen data")
        val brukerdefinert: Boolean
            get() = brukerutfyltVerdi != null || harIkkeKonto
    }

    private fun loadKontonummer(behandlingsId: String): JsonKontonummer =
        soknadUnderArbeidRepository.hentSoknad(behandlingsId, eier())
            .jsonInternalSoknad!!.soknad.data.personalia.kontonummer

    private fun storeKontonummer(
        behandlingsId: String,
        kontonummer: JsonKontonummer,
    ) {
        val soknad = soknadUnderArbeidRepository.hentSoknad(behandlingsId, eier())
        soknad.jsonInternalSoknad!!.soknad.data.personalia.kontonummer = kontonummer
        soknadUnderArbeidRepository.oppdaterSoknadsdata(soknad, eier())
    }

    private fun mapInputToJson(
        inputDto: KontonummerInputDto,
        kontoFraSystem: String?,
    ): JsonKontonummer {
        val definedByUser = inputDto.brukerutfyltVerdi != null || inputDto.harIkkeKonto == true

        return when (definedByUser) {
            true ->
                JsonKontonummer()
                    .withKilde(JsonKilde.BRUKER)
                    .withVerdi(inputDto.brukerutfyltVerdi)
                    .withHarIkkeKonto(inputDto.harIkkeKonto ?: false)

            false ->
                JsonKontonummer()
                    .withKilde(JsonKilde.SYSTEM)
                    .withVerdi(kontoFraSystem)
                    .withHarIkkeKonto(null)
        }
    }

    /**
     *  Jeg gleder meg til ny datamodell...
     *  DAO har ikke plass til både brukerutfyltVerdi og systemverdi,
     *  så dersom brukeren velger den ene, blir den andre slettet. ¯\_(ツ)_/¯
     */
    private fun mapJsonToDto(kontonummer: JsonKontonummer) =
        KontonummerFrontend(
            systemverdi = kontonummerService.getKontonummer(eier()),
            brukerutfyltVerdi = kontonummer.verdi.takeIf { kontonummer.kilde == JsonKilde.BRUKER },
            harIkkeKonto = kontonummer.harIkkeKonto ?: false,
        )
}

data class KontonummerInputDto(
    @Schema(nullable = true, description = "Kontonummer fra bruker")
    @field:Pattern(regexp = "^\\d{11}$", message = "Kontonummer må være 11 siffer")
    val brukerutfyltVerdi: String? = null,
    @Schema(nullable = true, description = "Bruker oppgir at de ikke har konto")
    val harIkkeKonto: Boolean? = null,
    @Deprecated("Ignorert - kun her for bakoverkompatibilitet")
    val brukerdefinert: Boolean? = null,
)
